### PR TITLE
[multi-asic] show_transceiver changes

### PR DESCRIPTION
**Pull request by judyjoseph**
**- What I did**
Show transceiver changes 
Rearranging of common functions in config/main.py to py-common, utilities_common
Dependent on another PR (https://github.com/Azure/sonic-buildimage/pull/5269)

**- How I did it**

**- How to verify it**
```
admin@str--acs-1:~$ sudo show interfaces transceiver eeprom 
Ethernet0: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031430

Ethernet4: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031360
.....

admin@str--acs-1:~$ sudo show interfaces transceiver eeprom -d
Ethernet0: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031430
        ChannelMonitorValues: 
                RX1Power: -infdBm
                RX2Power: -infdBm
                RX3Power: -infdBm
                RX4Power: -infdBm
                TX1Bias: 0.0000mA
                TX2Bias: 0.0000mA
                TX3Bias: 0.0000mA
                TX4Bias: 0.0000mA
        ChannelThresholdValues: 
                RxPowerHighAlarm  : 8.1647dBm
                RxPowerHighWarning: 8.1647dBm
                RxPowerLowAlarm   : 8.1647dBm
                RxPowerLowWarning : 8.1647dBm
                TxBiasHighAlarm   : 131.0700mA
                TxBiasHighWarning : 131.0700mA
                TxBiasLowAlarm    : 131.0700mA
                TxBiasLowWarning  : 131.0700mA
        ModuleMonitorValues: 
                Temperature: 1.5039C
                Vcc: 0.0000Volts
        ModuleThresholdValues: 
                TempHighAlarm  : -0.0039C
                TempHighWarning: -0.0039C
                TempLowAlarm   : -0.0039C
                TempLowWarning : -0.0039C
                VccHighAlarm   : 6.5535Volts
                VccHighWarning : 6.5535Volts
                VccLowAlarm    : 6.5535Volts
                VccLowWarning  : 6.5535Volts
.....
admin@str--acs-1:~$ sudo show interfaces transceiver eeprom -n asic0
Ethernet0: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031430

Ethernet4: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031360

Ethernet8: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2018-10-27 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110402104
        Vendor Rev: 
        Vendor SN: 1830031371


admin@str--acs-1:~$ sudo show interfaces transceiver presence --help
Usage: show interfaces transceiver presence [OPTIONS] [INTERFACENAME]

  Show interface transceiver presence

Options:
  -n, --namespace [asic0|asic1|asic2|asic3|asic4|asic5]
                                  Namespace name or all
  --verbose                       Enable verbose output
  -?, -h, --help                  Show this message and exit.
admin@str--acs-1:~$ sudo show interfaces transceiver presence 
Port         Presence
-----------  -----------
Ethernet0    Present
Ethernet4    Present
Ethernet8    Present
Ethernet12   Present
Ethernet16   Present
Ethernet20   Present
Ethernet24   Present
Ethernet28   Present
Ethernet32   Present
Ethernet36   Present
Ethernet40   Present
Ethernet44   Present
Ethernet48   Present
Ethernet52   Present
Ethernet56   Present
Ethernet60   Present
Ethernet64   Present
Ethernet68   Present
Ethernet72   Present
Ethernet76   Present
Ethernet80   Present
Ethernet84   Present
Ethernet88   Present
Ethernet92   Present
Ethernet96   Present
Ethernet100  Present
Ethernet104  Present
Ethernet108  Present
Ethernet112  Present
Ethernet116  Not present
Ethernet120  Present
Ethernet124  Present
Ethernet128  Present
Ethernet132  Present
Ethernet136  Present
Ethernet140  Present
Ethernet144  Present
Ethernet148  Present
Ethernet152  Present
Ethernet156  Present
Ethernet160  Present
Ethernet164  Present
Ethernet168  Present
Ethernet172  Present
Ethernet176  Present
Ethernet180  Present
Ethernet184  Present
Ethernet188  Present
Ethernet192  Present
Ethernet196  Present
Ethernet200  Present
Ethernet204  Present
Ethernet208  Present
Ethernet212  Present
Ethernet216  Present
Ethernet220  Present
Ethernet224  Present
Ethernet228  Present
Ethernet232  Present
Ethernet236  Not present
Ethernet240  Present
Ethernet244  Present
Ethernet248  Present
Ethernet252  Present
admin@str--acs-1:~$ sudo show interfaces transceiver presence -n asic0
Port        Presence
----------  ----------
Ethernet0   Present
Ethernet4   Present
Ethernet8   Present
Ethernet12  Present
Ethernet16  Present
Ethernet20  Present
Ethernet24  Present
Ethernet28  Present
Ethernet32  Present
Ethernet36  Present
Ethernet40  Present
Ethernet44  Present
Ethernet48  Present
Ethernet52  Present
Ethernet56  Present
Ethernet60  Present

admin@str--acs-1:~$ sudo show interfaces transceiver lpmode --help
Usage: show interfaces transceiver lpmode [OPTIONS] [INTERFACENAME]

  Show interface transceiver low-power mode status

Options:
  --verbose       Enable verbose output
  -?, -h, --help  Show this message and exit.
admin@str--acs-1:~$ sudo show interfaces transceiver lpmode
Port         Low-power Mode
-----------  ----------------
Ethernet0    On
Ethernet4    Off
Ethernet8    Off
Ethernet12   Off
Ethernet16   Off
Ethernet20   Off
Ethernet24   Off
Ethernet28   Off
Ethernet32   Off
Ethernet36   Off
Ethernet40   Off
Ethernet44   Off
Ethernet48   Off
Ethernet52   Off
Ethernet56   Off
Ethernet60   Off
Ethernet64   Off
Ethernet68   Off
Ethernet72   Off
Ethernet76   Off
Ethernet80   Off
Ethernet84   Off
Ethernet88   Off
Ethernet92   Off
Ethernet96   Off
Ethernet100  Off
Ethernet104  Off
Ethernet108  Off
Ethernet112  Off
Ethernet116  Off
Ethernet120  Off
Ethernet124  Off
Ethernet128  Off
Ethernet132  Off
Ethernet136  Off
Ethernet140  Off
Ethernet144  Off
Ethernet148  Off
Ethernet152  Off
Ethernet156  Off
Ethernet160  Off
Ethernet164  Off
Ethernet168  Off
Ethernet172  Off
Ethernet176  Off
Ethernet180  Off
Ethernet184  Off
Ethernet188  Off
Ethernet192  Off
Ethernet196  Off
Ethernet200  Off
Ethernet204  Off
Ethernet208  Off
Ethernet212  Off
Ethernet216  Off
Ethernet220  Off
Ethernet224  Off
Ethernet228  Off
Ethernet232  Off
Ethernet236  Off
Ethernet240  Off
Ethernet244  Off
Ethernet248  Off
Ethernet252  Off
```

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

**Comment by judyjoseph** [comment]
retest this please

File: scripts/sfpshow [hunk 1]
```diff
@@ -350,16 +366,34 @@ def cli():
...
+    sfp.get_eeprom()
+    sfp.display_eeprom()
 
 # 'presence' subcommand
 @cli.command()
 @click.option('-p', '--port', metavar='<port_name>', help="Display SFP presence for port <port_name> only")
-def presence(port):
-    sfp = SFPShow()
-    sfp.display_presence(port)
+@click.option('-n', '--namespace', default=None, help="Display interfaces for specific namespace")
```
Comment (arlakshm) [inline scripts/sfpshow:385]: can we use the click.choices in the commands in this file  like the commands below ? It will be consistent.

File: tests/mock_tables/mock_multi_asic.py [hunk 2]
```diff
@@ -14,7 +13,10 @@ def mock_is_multi_asic():
 def mock_get_namespace_list(namespace=None):
     return ['asic0', 'asic1']
 
+def mock_get_namespace_for_port(port):
```
Comment (arlakshm) [inline tests/mock_tables/mock_multi_asic.py:16]: Why to we need to mock get_namespace_for_port. I think if we mock the get_namespace_list, get_namespace_for_port API should work. Any issues seen ?

File: tests/sfp_test.py [hunk 3]
```diff
@@ -122,8 +122,88 @@ def test_sfp_eeprom(self):
         expected = "Ethernet200: SFP EEPROM Not detected"
         assert result_lines == expected
 
+    def test_sfp_presence_with_ns(self):
```
Comment (arlakshm) [inline tests/sfp_test.py:125]: The command has asic name when testing single asic . Is this a negative test case ?

File: tests/sfp_test.py [hunk 4]
```diff
@@ -122,8 +122,88 @@ def test_sfp_eeprom(self):
...
+
+class Test_multiAsic_SFP(object):
+    @classmethod
+    def setup_class(cls):
+        print("SETUP")
+        os.environ["PATH"] += os.pathsep + scripts_path
+        os.environ["UTILITIES_UNIT_TESTING"] = "2"
+        os.environ["UTILITIES_UNIT_TESTING_TOPOLOGY"] = "multi_asic"
+
+    def test_sfp_presence_with_ns(self):
```
Comment (arlakshm) [inline tests/sfp_test.py:158]: Do we need a test to display all interfaces from all namespaces ?

File: utilities_common/multi_asic.py [hunk 5]
```diff
@@ -5,7 +5,6 @@
 from sonic_py_common import multi_asic
 from utilities_common import constants
 
-
```
Comment (arlakshm) [inline utilities_common/multi_asic.py:8]: why this change ?

**Comment by arlakshm** [comment]
Do we need to support the lpmode commands as well ?

File: scripts/sfpshow [hunk 6]
```diff
@@ -291,53 +298,62 @@ class SFPShow(object):
...
-                out_put = self.convert_interface_sfp_info_to_cli_output_string(self.sdb, interfacename, dump_dom)
+                out_put = self.convert_interface_sfp_info_to_cli_output_string(self.db, self.intf_name, self.dump_dom)
             else:
-               out_put = out_put + interfacename + ': ' + 'SFP EEPROM Not detected' + '\n'
+               out_put = out_put + self.intf_name + ': ' + 'SFP EEPROM Not detected' + '\n'
         else:
-            port_table_keys = self.adb.keys(self.adb.APPL_DB, "PORT_TABLE:*")
+            port_table_keys = self.db.keys(self.db.APPL_DB, "PORT_TABLE:*")
             sorted_table_keys = natsorted(port_table_keys)
+            print(self.multi_asic.current_namespace)
```
[the remark this answers is not in the document]
Comment (judyjoseph) [reply scripts/sfpshow]: removed this print

File: scripts/sfpshow (hunk 1 above)
Comment (judyjoseph) [reply scripts/sfpshow:393]: click choice() is used in the CLI command handler in show/interfaces/__init.py__. So the user using CLI will get the choice of namespaces. It is not used in the sfpshow script as it was causing issues with the tests.

File: tests/mock_tables/mock_multi_asic.py (hunk 2 above)
Comment (judyjoseph) [reply tests/mock_tables/mock_multi_asic.py]: I have removed the same , this mock is not needed.

File: tests/sfp_test.py (hunk 3 above)
Comment (judyjoseph) [reply tests/sfp_test.py]: removed, this was my initial idea to club both single/multi asic in a single testcase. Later made a new multi-ASIC testcase but missed to remove this. thanks for pointing out.

File: tests/sfp_test.py (hunk 4 above)
Comment (judyjoseph) [reply tests/sfp_test.py:288]: added testcases for the same

**Comment by judyjoseph** [comment]
retest this please

**Review by arlakshm — APPROVED** [review]
lgtm

**Comment by abdosi** [comment]
@judyjoseph Please create PR for 201911.

**Comment by judyjoseph** [comment]
> @judyjoseph Please create PR for 201911.

It is already taken care of by this PR https://github.com/Azure/sonic-utilities/pull/1087 ( as there were some dependencies )

File: tests/sfp_test.py [hunk 7]
```diff
@@ -127,3 +275,83 @@ def teardown_class(cls):
...
+        runner = CliRunner()
+        result = runner.invoke(show.cli.commands["interfaces"].commands["transceiver"].commands["presence"], ["Ethernet0 -n asic0"])
```
Comment (maipbui) [inline tests/sfp_test.py:290]: >["Ethernet0 -n asic0"] [](http://example.com/codeflow?start=109&length=22)

I think this is incorrect way to parse arguments. Ref: [https://click.palletsprojects.com/en/8.1.x/testing/#id4](https://click.palletsprojects.com/en/8.1.x/testing/#id4)

File: tests/sfp_test.py [hunk 8]
```diff
@@ -127,3 +275,83 @@ def teardown_class(cls):
...
+        result = runner.invoke(show.cli.commands["interfaces"].commands["transceiver"].commands["eeprom"], ["Ethernet0 -n asic0"])
+        assert result.exit_code == 0
+        assert "\n".join([ l.rstrip() for l in result.output.split('\n')]) == test_sfp_eeprom_output
+
+        result = runner.invoke(show.cli.commands["interfaces"].commands["transceiver"].commands["eeprom"], ["Ethernet200 -n asic0"])
+        result_lines = result.output.strip('\n')
+        expected = "Ethernet200: SFP EEPROM Not detected"
+        assert result_lines == expected
+
+    def test_sfp_eeprom_with_ns(self):
```
Comment (maipbui) [inline tests/sfp_test.py:329]: >def test_sfp_eeprom_with_ns(self): [](http://example.com/codeflow?start=4&length=34)

duplicate UT
